marlowe-dashboard-client/src/MainFrame/State.purs

```diff
@@ -109,13 +111,44 @@ handleQuery (ReceiveWebSocketMessage msg next) = do
     (WS.WebSocketClosed reason) -> assign _webSocketStatus (WebSocketClosed (Just reason)) -- TODO: show warning to the user
     (WS.ReceiveMessage (Left errors)) -> pure unit -- TODO: show error to the user
     (WS.ReceiveMessage (Right streamToClient)) -> case streamToClient of
-      InstanceUpdate contractInstanceId instanceStatusToClient -> pure unit
+      -- update the current slot
       SlotChange slot -> assign (_playState <<< _currentSlot) $ toFront slot
+      -- update the wallet funds (if the change is to the current wallet)
+      -- note: we should only ever be notified of changes to the current wallet,
+      -- since we subscribe to this update when we pick it up, and unsubscribe
+      -- when we put it down - but we check here anyway in case
       WalletFundsChange wallet value -> do
         mCurrentWallet <- peruse (_playState <<< _walletDetails <<< _wallet)
         for_ mCurrentWallet \currentWallet ->
           when (currentWallet == toFront wallet)
             $ assign (_playState <<< _walletDetails <<< _assets) (toFront value)
+      -- update the state when a contract instance changes
+      -- note: we should be subsribed to updates from all (and only) the current
+      -- wallet's contract instances, including its wallet companion contract
+      InstanceUpdate bContractInstanceId instanceStatusToClient -> case instanceStatusToClient of
+        NewObservableState rawJson -> do
+          let
+            contractInstanceId = toFront bContractInstanceId
+          mPlayState <- peruse _playState
```

Comment:
Do we need to handle this here or can we do it in the PlayState? I ask so that we don't need to do the `for _mPlayState`

Reply:
It would be better in the Play.State, but I wasn't immediately sure the best way to do that (without making Play a genuine child component with it's own query handler). Maybe just have an action in the Play.State, and call that action here?